test/e2e/node_feature_discovery_test.go

```diff
@@ -1018,6 +1017,305 @@ resyncPeriod: "1s"
 				Expect(err).NotTo(HaveOccurred())
 			})
 		})
+
+		Context("selected namespaces restriction is respected or not", func() {
+			BeforeEach(func(ctx context.Context) {
+				extraMasterPodSpecOpts = []testpod.SpecOption{
+					testpod.SpecWithConfigMap("nfd-master-conf", "/etc/kubernetes/node-feature-discovery"),
+				}
+				cm := testutils.NewConfigMap("nfd-master-conf", "nfd-master.conf", `
+restrictions:
+  nodeFeatureNamespaceSelector:
+    matchLabels:
+      kubernetes.io/metadata.name: "fake"
+`)
+				_, err := f.ClientSet.CoreV1().ConfigMaps(f.Namespace.Name).Create(ctx, cm, metav1.CreateOptions{})
+				Expect(err).NotTo(HaveOccurred())
+			})
+			It("Nothing should be created", func(ctx context.Context) {
+				// deploy node feature object
+				nodes, err := getNonControlPlaneNodes(ctx, f.ClientSet)
+				Expect(err).NotTo(HaveOccurred())
+
+				targetNodeName := nodes[0].Name
+				Expect(targetNodeName).ToNot(BeEmpty(), "No suitable worker node found")
+
+				// Apply Node Feature object
+				By("Creating NodeFeature object")
+				nodeFeatures, err := testutils.CreateOrUpdateNodeFeaturesFromFile(ctx, nfdClient, "nodefeature-1.yaml", f.Namespace.Name, targetNodeName)
+				Expect(err).NotTo(HaveOccurred())
+
+				By("Verifying node labels from NodeFeature object #1 are not created")
+				// No labels should be created since the f.Namespace is not in the selected Namespaces
+				expectedLabels := map[string]k8sLabels{
+					targetNodeName: {},
+					"*":            {},
+				}
+				eventuallyNonControlPlaneNodes(ctx, f.ClientSet).Should(MatchLabels(expectedLabels, nodes))
```

Comment:
The issue is with the last verification because when a NodeFeature is not selected in the current implementation, we don't do anything. We don't act as if it was deleted, therefore labels won't be removed even tho the NodeFeature is not selected anymore.

Reply:
We can do it the other way around, were start from an initial state where we have no labels, we create a non-selected NodeFeature, and then we patch the namespace with specific labels to make NodeFeature selectable and therefore we can see labels. WDYT?

Reply:
> The issue is with the last verification because when a NodeFeature is not selected in the current implementation, we don't do anything. We don't act as if it was deleted, therefore labels won't be removed even tho the NodeFeature is not selected anymore.

Yup, you're right. That's why I suggested running nfd-master with a resync-period of 1s -> forces full evaluation of CRDs every second. When the NodeFeature gets "un-selected" the related labels should be removed.



> We can do it the other way around, were start from an initial state where we have no labels, we create a non-selected NodeFeature, and then we patch the namespace with specific labels to make NodeFeature selectable and therefore we can see labels. WDYT?

This also needs the quick resync-period in order to mitigate the "we do nothing" problem. But again, I think this has the same raciness problem, hard to be entirely sure when the un-selected NF was processed. So I think the way I suggested is safer because you have clear triggers to watch when nfd-master acted. WDYT?

Reply:
> When the NodeFeature gets "un-selected" the related labels should be removed.

This is the part missing in the implementation. If you check the current code, if it's unselected we simply log that it's not selected and we continue. There's no removing the labels in that case. Do you think that it should be added? 
I think filtering of namespaces should happen at the level of  `getAndMergeNodeFeatures`  function. thoughts?

Reply:
Ach, exactly :man_facepalming: That is a key part missing. We really need to do the filtering there (too). If we only do the filtering in the event functions it doesn't prevent nfd-master to using all the NFs